File: Python3/Spike-contrast.py
# -*- coding: utf-8 -*-
"""
Created on Tue Apr 25 17:59:39 2017

@author: Manuel Ciba
"""

import numpy as np
import time

def generateRandomTestData(num_trains, num_spikes, T):
    spike_trains = np.random.random((num_trains, num_spikes))   # random values between 0 and 1
    spike_trains = np.sort(spike_trains)                        # sort values
    spike_trains = spike_trains * T
    return spike_trains

def generateTestData():                       
    spike_trains = np.array([[1, 1 ,1], [2, 3, 1.5], [5, 8, 6], [9, 10, 9]])
    T = 11
    return spike_trains, T

def get_Theta_and_n_perBin(spike_trains, T, binSize, binStepFactor):
   
    if binStepFactor == 2:
        # init
        binStep = binSize / binStepFactor
        edges = np.arange(0, T+binStep, binStep)
        tmp = spike_trains.shape                                    # number of spike trains
        N = tmp[1]
        hist = np.zeros((len(edges)-1,N));
        # calculate histogram for every spike train
        for i in range(0, N):                                       # for all spike trains
            hist[:,i] = binning_halfOverlap(spike_trains[:,i], T, binSize)
        # calculate parameter over all spike trains
        theta = np.sum(hist, 1)                                      # number of spikes per bin 
        mask = hist != 0
        n = np.sum(mask, 1)                                          # number of active spike trains
        
    return theta, n

def binning_halfOverlap(y, T, binSize, flag_binary=False):
    binStep = binSize/2
    edges = np.arange(0, T+binStep, binStep)
    hist, bin_edges = np.histogram(y, edges)
    hist[0:len(hist)-1] = hist[0:len(hist)-1] + hist[1:len(hist)] 
    return hist

"""
Main Function: SpikeContrast
"""

def SpikeContrast(spike_trains, T):
    # parameter
    binShrinkFactor = 0.9                                       # bin size decreases by factor 0.9 for each iteration
    binStepFactor = 2                                           # bin overlap of 50 %
    binMax = T/2
    isi = np.diff(spike_trains, axis=0)
    isiMin = np.min(isi)
    binMin = np.max([isiMin, 0.01])
    
    tmp = spike_trains.shape                                    # number of spike trains
    N = tmp[1]
    
    # initialization
    numIterations = np.ceil(np.log(binMin / binMax) / np.log(binShrinkFactor))
    numIterations = int(numIterations)
    ActiveST = np.zeros((numIterations, 1))
    Contrast = np.zeros((numIterations, 1))
    C = np.zeros((numIterations, 1))
    
    numAllSpikes = spike_trains.size
    binSize = binMax
    
    for i in range(0, numIterations):                           # for 0, 1, 2, ... numIterations
             
        # calculate Theta and n
        Theta_k, n_k = get_Theta_and_n_perBin(spike_trains, T, binSize, binStepFactor)
        # calcuate C= Contrast * ActiveST
        ActiveST[i] = ((np.sum(n_k*Theta_k))/(np.sum(Theta_k))-1) / (N-1)
        Contrast[i] = (np.sum(np.abs(np.diff(Theta_k)))/ (numAllSpikes*2))                    # Contrast: sum(|derivation|) / (2*#Spikes)
        C[i] = Contrast[i] * ActiveST[i]                                                      # Synchrony curve "C"
        
        binSize *= binShrinkFactor                              # new bin size
    
    # Sync value is maximum of cost function C
    S = np.max(C)
    return S


"""
Main Program
"""

spike_trains, T = generateTestData()
#spike_trains = generateRandomTestData(3, 5, 60)
S = SpikeContrast(spike_trains, T)


T = 100
spike_trains = generateRandomTestData(1000,1000,T)
start = time.time()
S = SpikeContrast(spike_trains, T)
end = time.time()
print(end - start)

File: MATLAB/SpikeContrast.m
%% Spike-contrast 
% Purpose:  Measure synchrony between two or more spike trains
%
% Input:    TS:         time stamps of spikes in seconds, 
%                       input as matrix (dim: maxNumOfSpikesPerSpiketrain x numOfSpiketrains)
%                       (NaN-Padding)
%           T:          signal length in seconds
%
% Output:   S:          Synchrony index
%           PREF:       Used preferences/parameter to calculate Spike-contrast
%
% needed functions:     [AllSpikesPerBin,actElPerBin,edges,xvalues]=getNumSpikesAndActElPerBin(TS,rec_dur,bin);
%                       [y_binned,x_step,edges_step]=binning(y,rec_dur,binsize,step,flag_binary) 
%
% written by Manuel Ciba, 2016/2017



function [S,PREF] = SpikeContrast(TS,T)
    
    tic

    %% FORMATTING
    TS(TS==0)=NaN;                          % force NaN patting (not correct if spike appear at zero seconds)
    TS=sort(TS);    
    if any(any(TS<0))
        error('negative time stamps are not allowed')
    end
    
    %% PARAMETER:
    % bin size decrease factor:
    binShrinkFactor=0.9;           
    % max. bin size:
    maxBin=T/2;                                              
    % min. bin size:
    ISI=diff(TS);
    ISImin= min(min(ISI));
    minBin=max([ISImin/2, 0.001]); % minBin is not smaller than 1 ms   
    minBin=0.001;
    % bin overlap:
    binStepFactor = 2;    

    if maxBin <= minBin
        error('min. bin size is greater or equal to max. bin size')
    end
    
    %% 0) N = number of spike trains
    N=sum(max(~isnan(TS))); % if no spikes on spike train, than every element is NaN

    %% 1) Binning of TS 

    % initialization:
    numIterations = ceil(log(minBin/maxBin)/log(binShrinkFactor)); % maxBin * binShrinkFactor^numIterations >= minBin
    ActiveST=zeros(numIterations,1); 
    Contrast=zeros(numIterations,1); 
    C=zeros(numIterations,1); % C = ActiveST * Contrast
    bins=zeros(numIterations,1);
    
    % calc all histograms
    numAllSpikes = sum(~isnan(TS(:))); % number of all spikes in TS
    bin = maxBin; % first bin size
    for i=1:numIterations

        step = bin/binStepFactor; % bin step
        
        %% Calculate Theta and n  
        time_start = -ISImin;
        time_end = T+ISImin;
        [Theta_k,n_k]=f_SC_get_Theta_and_n_perBin(TS,time_start,time_end,bin,step);
        
        %% Calculate: C = Contrast * ActiveST          
        ActiveST(i) = ((sum(n_k.*Theta_k))/(sum(Theta_k))-1) / (N-1);
        Contrast(i)=(sum(abs(diff(Theta_k)))/ (numAllSpikes*2)) ; % Contrast: sum(|derivation|) / (2*#Spikes)
        C(i)=Contrast(i) * ActiveST(i); % Synchrony curve "C"
        
        bins(i)=bin; % safe bins
        bin=bin*binShrinkFactor; % new decreased bin size

    end

    %% 2) Sync value is maximum of cost function C
    S= max(C);
    PREF.C = C;
    PREF.Contrast = Contrast;
    PREF.ActiveST = ActiveST;
    PREF.bins = bins;
    PREF.maxBin = maxBin;
    PREF.minBin = minBin;
    PREF.binStepFactor = binStepFactor;
    PREF.binShrinkFactor = binShrinkFactor;
    PREF.T = T;
    PREF.t = toc; % calculation time
    
end